src/coreclr/pal/src/exception/signal.cpp

```diff
@@ -838,21 +838,53 @@ static void inject_activation_handler(int code, siginfo_t *siginfo, void *contex
 Function :
     InjectActivationInternal
 
-    Interrupt the specified thread and have it call the activationFunction passed in
+    Interrupt the specified thread with INJECT_ACTIVATION_SIGNAL
 
 Parameters :
     pThread            - target PAL thread
-    activationFunction - function to call
 
-(no return value)
+Return:
+   PAL_ERROR
 --*/
 PAL_ERROR InjectActivationInternal(CorUnix::CPalThread* pThread)
 {
 #ifdef INJECT_ACTIVATION_SIGNAL
-    int status = pthread_kill(pThread->GetPThreadSelf(), INJECT_ACTIVATION_SIGNAL);
-    // We can get EAGAIN when printing stack overflow stack trace and when other threads hit
-    // stack overflow too. Those are held in the sigsegv_handler with blocked signals until
-    // the process exits.
+    int status = 0;
+
+#ifdef __linux__
+    const int signal_queue_ovf_retry_count = 5;
+    struct timespec signal_queue_ovf_sleep;
+    signal_queue_ovf_sleep.tv_sec = 0;
+    signal_queue_ovf_sleep.tv_nsec = 10 * 1000 * 1000; /* 10 milliseconds */
+    int retry_count = 0;
+
+    while (true)
+#endif // __linux__
+    {
+        status = pthread_kill(pThread->GetPThreadSelf(), INJECT_ACTIVATION_SIGNAL);
+        // We can get EAGAIN when printing stack overflow stack trace and when other threads hit
+        // stack overflow too. Those are held in the sigsegv_handler with blocked signals until
+        // the process exits.
+        //
+        // We can also get EAGAIN if the signal queue overflows on linux.
+
+#ifdef __linux__
+        if (status == EAGAIN && retry_count < signal_queue_ovf_retry_count)
```

Comment:
I do not think we need to have a retry loop here. The caller of `InjectActivation` in the runtime has a retry loop already. I think we can just return failure from here - similar to how we return `ENOTSUP` on Apple - and let the runtime retry.

Reply:
Or maybe we should just delete `PROCAbort` on failure and propagate all error to runtime.

@janvorli Thoughts?

Reply:
We never check the return code of the `InjectActivation` for other purposes than logging a stresslog entry.

Reply:
We rely on the retry loop in the runtime. The `InjectActivation` can succeed and still the callback would never be called if the thread was executing a native code.